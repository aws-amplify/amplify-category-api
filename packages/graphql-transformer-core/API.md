## API Report File for "graphql-transformer-core"

> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).

```ts

/// <reference types="node" />

import { Condition } from 'cloudform-types/types/dataTypes';
import { ConditionIntrinsicFunction } from 'cloudform-types';
import { default as default_2 } from 'cloudform-types/types/dynamoDb/table';
import { default as default_3 } from 'cloudform-types/types/iam/role';
import { Diff as Diff_2 } from 'deep-diff';
import { DirectiveDefinitionNode } from 'graphql';
import { DirectiveNode } from 'graphql';
import { DocumentNode } from 'graphql';
import { EnumTypeDefinitionNode } from 'graphql';
import { EnumTypeExtensionNode } from 'graphql/language/ast';
import { EnumValueDefinitionNode } from 'graphql';
import { FieldDefinitionNode } from 'graphql';
import { GraphQLError } from 'graphql';
import { InputObjectTypeDefinitionNode } from 'graphql';
import { InputObjectTypeExtensionNode } from 'graphql/language/ast';
import { InputValueDefinitionNode } from 'graphql';
import { InterfaceTypeDefinitionNode } from 'graphql';
import { InterfaceTypeExtensionNode } from 'graphql/language/ast';
import { ObjectTypeDefinitionNode } from 'graphql';
import { ObjectTypeExtensionNode } from 'graphql';
import Output from 'cloudform-types/types/output';
import Parameter from 'cloudform-types/types/parameter';
import { Policy } from 'cloudform-types/types/iam/role';
import { Readable } from 'stream';
import Resource from 'cloudform-types/types/resource';
import { ScalarTypeDefinitionNode } from 'graphql';
import { SchemaDefinitionNode } from 'graphql';
import Template from 'cloudform-types/types/template';
import { Template as Template_2 } from 'cloudform-types';
import { TypeDefinitionNode } from 'graphql';
import { TypeSystemDefinitionNode } from 'graphql';
import { UnionTypeDefinitionNode } from 'graphql';
import { UnionTypeExtensionNode } from 'graphql/language/ast';

// Warning: (ae-forgotten-export) The symbol "ProjectOptions" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export function buildAPIProject(opts: ProjectOptions): Promise<{
    resolvers: StringMap;
    stacks: {
        [name: string]: Template_2;
    };
    stackMapping: StackMapping_3;
    pipelineFunctions: StringMap;
    functions: {
        [path: string]: string;
    };
    schema: string;
    rootStack: Template_2;
}>;

// Warning: (ae-forgotten-export) The symbol "DiffableProject" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export const cantAddAndRemoveGSIAtSameTimeRule: (diff: Diff_2<DiffableProject, DiffableProject>, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export const cantBatchMutateGSIAtUpdateTimeRule: (diff: Diff_2<DiffableProject, DiffableProject>, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export const cantEditGSIKeySchemaRule: (diff: Diff_2<DiffableProject, DiffableProject>, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// Warning: (ae-forgotten-export) The symbol "Diff" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export const cantHaveMoreThan500ResourcesRule: (diffs: Diff[], currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export const cantMutateMultipleGSIAtUpdateTimeRule: (diffs: Diff[], currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export const cantRemoveTableAfterCreation: (_: Diff_2<DiffableProject, DiffableProject>, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export const CLOUDFORMATION_FILE_NAME = "cloudformation-template.json";

// @public (undocumented)
export function collectDirectiveNames(sdl: string): string[];

// @public (undocumented)
export function collectDirectivesByType(sdl: string): Object;

// @public (undocumented)
export function collectDirectivesByTypeNames(sdl: string): {
    types: Object;
    directives: string[];
};

// @public (undocumented)
export type ConflictDetectionType = 'VERSION' | 'NONE';

// @public (undocumented)
export const enum ConflictHandlerType {
    // (undocumented)
    AUTOMERGE = "AUTOMERGE",
    // (undocumented)
    LAMBDA = "LAMBDA",
    // (undocumented)
    OPTIMISTIC = "OPTIMISTIC_CONCURRENCY"
}

// Warning: (ae-forgotten-export) The symbol "ResolversFunctionsAndSchema" needs to be exported by the entry point index.d.ts
// Warning: (ae-forgotten-export) The symbol "NestedStacks" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export interface DeploymentResources extends ResolversFunctionsAndSchema, NestedStacks {
    // (undocumented)
    stackMapping: StackMapping_3;
}

// @public (undocumented)
export class DestructiveMigrationError extends Error {
    constructor(message: string, removedModels: string[], replacedModels: string[]);
    // (undocumented)
    toString: () => string;
}

// @public (undocumented)
export type DiffRule = (diff: Diff, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export interface FeatureFlagProvider {
    // (undocumented)
    getBoolean(featureName: string, defaultValue?: boolean | null): boolean;
    // (undocumented)
    getNumber(featureName: string, defaultValue?: Number): Number;
    // (undocumented)
    getObject(featureName: string, defaultValue?: object): object;
}

// @public (undocumented)
export function getAppSyncServiceExtraDirectives(): string;

// @public (undocumented)
export const getCantAddLSILaterRule: (iterativeUpdatesEnabled?: boolean) => (diff: Diff_2<DiffableProject, DiffableProject>) => void;

// @public (undocumented)
export const getCantEditKeySchemaRule: (iterativeUpdatesEnabled?: boolean) => (diff: Diff_2<DiffableProject, DiffableProject>) => void;

// @public (undocumented)
export const getCantEditLSIKeySchemaRule: (iterativeUpdatesEnabled?: boolean) => (diff: Diff_2<DiffableProject, DiffableProject>, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export const getCantRemoveLSILater: (iterativeUpdatesEnabled?: boolean) => (diff: Diff_2<DiffableProject, DiffableProject>, currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
export function getDirectiveArguments(directive: DirectiveNode): any;

// @public (undocumented)
export function getFieldArguments(type: any): any;

// @public (undocumented)
export function getSanityCheckRules(isNewAppSyncAPI: boolean, ff: FeatureFlagProvider, allowDestructiveUpdates?: boolean): {
    diffRules: DiffRule[];
    projectRules: ProjectRule[];
};

// @public (undocumented)
export function getTableNameForModel(sdl: string, modelName: string): string;

// @public (undocumented)
export function gql(literals: TemplateStringsArray, ...placeholders: string[]): DocumentNode;

// @public (undocumented)
export class GraphQLTransform {
    // Warning: (ae-forgotten-export) The symbol "GraphQLTransformOptions" needs to be exported by the entry point index.d.ts
    constructor(options: GraphQLTransformOptions);
    // (undocumented)
    transform(schema: string): DeploymentResources;
}

// @public (undocumented)
export class InvalidDirectiveError extends Error {
    constructor(message: string);
}

// @public (undocumented)
export class InvalidGSIMigrationError extends InvalidMigrationError {
    constructor(message: string, causedBy: string, fix: string);
    // (undocumented)
    causedBy: string;
    // (undocumented)
    fix: string;
}

// @public (undocumented)
export class InvalidMigrationError extends Error {
    constructor(message: string, causedBy: string, fix: string);
    // (undocumented)
    causedBy: string;
    // (undocumented)
    fix: string;
    // (undocumented)
    toString: () => string;
}

// @public (undocumented)
export class InvalidTransformerError extends Error {
    constructor(message: string);
}

// @public (undocumented)
export const isDataStoreEnabled: (projectDir: string) => Promise<boolean>;

// @public (undocumented)
export interface ITransformer {
    // (undocumented)
    after?: (acc: TransformerContext) => void;
    // (undocumented)
    argument?: (definition: InputValueDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    before?: (acc: TransformerContext) => void;
    // (undocumented)
    directive: DirectiveDefinitionNode;
    // (undocumented)
    enum?: (definition: EnumTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    enumValue?: (definition: EnumValueDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    field?: (parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode, definition: FieldDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    input?: (definition: InputObjectTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    inputValue?: (definition: InputValueDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    interface?: (definition: InterfaceTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    name: string;
    // (undocumented)
    object?: (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    scalar?: (definition: ScalarTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    typeDefinitions: TypeDefinitionNode[];
    // (undocumented)
    union?: (definition: UnionTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
}

// @public (undocumented)
function loadConfig(projectDir: string): Promise<TransformConfig>;
export { loadConfig }
export { loadConfig as readTransformerConfiguration }

// Warning: (ae-forgotten-export) The symbol "ProjectConfiguration" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
function loadProject(projectDirectory: string, opts?: ProjectOptions): Promise<ProjectConfiguration>;
export { loadProject }
export { loadProject as readProjectConfiguration }

// @public (undocumented)
export interface MappingParameters {
    // (undocumented)
    [key: string]: {
        [key: string]: {
            [key: string]: string | number | string[];
        };
    };
}

// Warning: (ae-forgotten-export) The symbol "MigrationOptions" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export function migrateAPIProject(opts: MigrationOptions): Promise<{
    project: any;
    cloudBackend: any;
}>;

// @public (undocumented)
export const PARAMETERS_FILE_NAME = "parameters.json";

// @public (undocumented)
export type ProjectRule = (diffs: Diff[], currentBuild: DiffableProject, nextBuild: DiffableProject) => void;

// @public (undocumented)
function readSchema(projectDirectory: string): Promise<string>;
export { readSchema as readProjectSchema }
export { readSchema }

// @public (undocumented)
export type ResolverConfig = {
    project?: SyncConfig;
    models?: {
        [key: string]: SyncConfig;
    };
};

// Warning: (ae-forgotten-export) The symbol "AmplifyApiV1Project" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export function revertAPIMigration(directory: string, oldProject: AmplifyApiV1Project): Promise<void>;

// @public (undocumented)
export const sanityCheckDiffs: (diffs: Diff[], current: DiffableProject, next: DiffableProject, diffRules: DiffRule[], projectRules: ProjectRule[]) => void;

// @public (undocumented)
export const sanityCheckProject: (currentCloudBackendDir: string, buildDirectory: string, rootStackName: string, diffRules: DiffRule[], projectRule: ProjectRule[]) => Promise<void>;

// @public (undocumented)
export class SchemaValidationError extends Error {
    constructor(errors: Readonly<GraphQLError[]>);
}

// @public (undocumented)
export function stripDirectives(doc: DocumentNode, except?: string[]): DocumentNode;

// @public (undocumented)
export type SyncConfig = SyncConfigOPTIMISTIC | SyncConfigSERVER | SyncConfigLAMBDA;

// @public (undocumented)
export type SyncConfigLAMBDA = {
    ConflictDetection: ConflictDetectionType;
    ConflictHandler: ConflictHandlerType.LAMBDA;
    LambdaConflictHandler: {
        name: string;
        region?: string;
        lambdaArn?: any;
    };
};

// @public (undocumented)
export type SyncConfigOPTIMISTIC = {
    ConflictDetection: ConflictDetectionType;
    ConflictHandler: ConflictHandlerType.OPTIMISTIC;
};

// @public (undocumented)
export type SyncConfigSERVER = {
    ConflictDetection: ConflictDetectionType;
    ConflictHandler: ConflictHandlerType.AUTOMERGE;
};

// @public (undocumented)
export module SyncUtils {
    // (undocumented)
    export function createSyncIAMRole(): default_3;
    // (undocumented)
    export function createSyncLambdaIAMPolicy({ name, region }: {
        name: string;
        region?: string;
    }): Policy;
    // (undocumented)
    export function createSyncTable(): default_2;
    // (undocumented)
    export function isLambdaSyncConfig(obj: any): obj is SyncConfigLAMBDA;
    // (undocumented)
    export function lambdaArnKey(name: string, region?: string): string;
    // Warning: (ae-forgotten-export) The symbol "DeltaSyncConfig" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    export function syncDataSourceConfig(): DeltaSyncConfig;
    // (undocumented)
    export function syncLambdaArnResource({ name, region }: {
        name: string;
        region?: string;
    }): ConditionIntrinsicFunction;
    // (undocumented)
    export function syncLambdaIAMRole({ name, region }: {
        name: string;
        region?: string;
    }): default_3;
    // Warning: (ae-forgotten-export) The symbol "SyncConfig_2" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    export function syncResolverConfig(syncConfig: SyncConfigOPTIMISTIC | SyncConfigLAMBDA | SyncConfigSERVER): SyncConfig_2;
    // (undocumented)
    export function syncTTLConfig(): {
        AttributeName: string;
        Enabled: boolean;
    };
}

// @public (undocumented)
export const TRANSFORM_BASE_VERSION = 4;

// @public (undocumented)
export const TRANSFORM_CONFIG_FILE_NAME = "transform.conf.json";

// @public (undocumented)
export const TRANSFORM_CURRENT_VERSION = 5;

// @public (undocumented)
export interface TransformConfig {
    // (undocumented)
    ElasticsearchWarning?: boolean;
    // (undocumented)
    Migration?: TransformMigrationConfig;
    // (undocumented)
    ResolverConfig?: ResolverConfig;
    // (undocumented)
    StackMapping?: {
        [resourceId: string]: string;
    };
    // (undocumented)
    TransformerOptions?: {
        [transformer: string]: {
            [option: string]: any;
        };
    };
    // (undocumented)
    transformers?: string[];
    // (undocumented)
    Version?: number;
    // (undocumented)
    warningESMessage?: boolean;
}

// @public (undocumented)
class Transformer_2 implements ITransformer {
    constructor(name: string, document: DocumentNode | string);
    // (undocumented)
    after?: (acc: TransformerContext) => void;
    // (undocumented)
    argument?: (definition: InputValueDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    before?: (acc: TransformerContext) => void;
    // (undocumented)
    directive: DirectiveDefinitionNode;
    // (undocumented)
    enum?: (definition: EnumTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    enumValue?: (definition: EnumValueDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    field?: (parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode, definition: FieldDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    input?: (definition: InputObjectTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    inputValue?: (definition: InputValueDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    interface?: (definition: InterfaceTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    name: string;
    // (undocumented)
    object?: (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    scalar?: (definition: ScalarTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
    // (undocumented)
    typeDefinitions: TypeDefinitionNode[];
    // (undocumented)
    union?: (definition: UnionTypeDefinitionNode, directive: DirectiveNode, acc: TransformerContext) => void;
}
export { Transformer_2 as Transformer }

// @public (undocumented)
export class TransformerContext {
    constructor(inputSDL: string, featureFlags: FeatureFlagProvider);
    // (undocumented)
    addEnum(en: EnumTypeDefinitionNode): void;
    // (undocumented)
    addEnumExtension(obj: EnumTypeExtensionNode): void;
    // (undocumented)
    addInput(inp: InputObjectTypeDefinitionNode): void;
    // (undocumented)
    addInputExtension(obj: InputObjectTypeExtensionNode): void;
    // (undocumented)
    addInterfaceExtension(obj: InterfaceTypeExtensionNode): void;
    // (undocumented)
    addMutationFields(fields: FieldDefinitionNode[]): void;
    // (undocumented)
    addObject(obj: ObjectTypeDefinitionNode): void;
    // (undocumented)
    addObjectExtension(obj: ObjectTypeExtensionNode): void;
    // (undocumented)
    addQueryFields(fields: FieldDefinitionNode[]): void;
    // (undocumented)
    addSubscriptionFields(fields: FieldDefinitionNode[]): void;
    // (undocumented)
    addType(obj: TypeDefinitionNode): void;
    // (undocumented)
    addUnionExtension(obj: UnionTypeExtensionNode): void;
    // (undocumented)
    readonly featureFlags: FeatureFlagProvider;
    // (undocumented)
    getMutation(): ObjectTypeDefinitionNode | undefined;
    // (undocumented)
    getMutationTypeName(): string | undefined;
    // (undocumented)
    getObject(name: string): ObjectTypeDefinitionNode | undefined;
    // (undocumented)
    getOutput(key: string): Output;
    // (undocumented)
    getQuery(): ObjectTypeDefinitionNode | undefined;
    // (undocumented)
    getQueryTypeName(): string | undefined;
    // (undocumented)
    getResolverConfig(): ResolverConfig;
    // (undocumented)
    getResource(resource: string): Resource;
    // (undocumented)
    getSchema(): SchemaDefinitionNode;
    // Warning: (ae-forgotten-export) The symbol "StackMapping" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    getStackMapping(): StackMapping;
    // (undocumented)
    getSubscription(): ObjectTypeDefinitionNode | undefined;
    // (undocumented)
    getSubscriptionTypeName(): string | undefined;
    // (undocumented)
    getTransformerVersion(): Number;
    // (undocumented)
    getType(name: string): TypeSystemDefinitionNode | undefined;
    // (undocumented)
    getTypeDefinitionsOfKind(kind: string): TypeDefinitionNode[];
    // (undocumented)
    inputDocument: DocumentNode;
    // (undocumented)
    isProjectUsingDataStore(): boolean;
    // (undocumented)
    mapResourceToStack(stackName: string, resource: string): void;
    // (undocumented)
    mergeConditions(conditions: {
        [key: string]: Condition;
    }): void;
    // (undocumented)
    mergeMappings(mapping: MappingParameters): void;
    // (undocumented)
    mergeOutputs(outputs: {
        [key: string]: Output;
    }): void;
    // (undocumented)
    mergeParameters(params: {
        [key: string]: Parameter;
    }): void;
    // (undocumented)
    mergeResources(resources: {
        [key: string]: Resource;
    }): void;
    // Warning: (ae-forgotten-export) The symbol "TransformerContextMetadata" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    metadata: TransformerContextMetadata;
    // (undocumented)
    nodeMap: {
        [name: string]: TypeSystemDefinitionNode;
    };
    // (undocumented)
    putSchema(obj: SchemaDefinitionNode): void;
    // (undocumented)
    putType(obj: TypeDefinitionNode): void;
    // (undocumented)
    setOutput(key: string, output: Output): void;
    // (undocumented)
    setResolverConfig(resolverConfig: ResolverConfig): void;
    // (undocumented)
    setResource(key: string, resource: Resource): void;
    // (undocumented)
    setTransformerVersion(version: Number): void;
    // (undocumented)
    template: Template;
    // (undocumented)
    updateObject(obj: ObjectTypeDefinitionNode): void;
}

// @public (undocumented)
export class TransformerContractError extends Error {
    constructor(message: string);
}

// @public (undocumented)
export interface TransformMigrationConfig {
    // (undocumented)
    V1?: {
        Resources: string[];
    };
}

// @public (undocumented)
export class UnknownDirectiveError extends Error {
    constructor(message: string);
}

// Warning: (ae-forgotten-export) The symbol "UploadOptions" needs to be exported by the entry point index.d.ts
//
// @public (undocumented)
export function uploadAPIProject(opts: UploadOptions): Promise<void>;

// @public (undocumented)
function writeConfig(projectDir: string, config: TransformConfig): Promise<TransformConfig>;
export { writeConfig }
export { writeConfig as writeTransformerConfiguration }

// Warnings were encountered during analysis:
//
// src/util/amplifyUtils.ts:50:3 - (ae-forgotten-export) The symbol "StringMap" needs to be exported by the entry point index.d.ts
// src/util/amplifyUtils.ts:54:3 - (ae-forgotten-export) The symbol "StackMapping_3" needs to be exported by the entry point index.d.ts

// (No @packageDocumentation comment for this package)

```
