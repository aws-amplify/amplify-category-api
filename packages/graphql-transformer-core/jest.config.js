const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
