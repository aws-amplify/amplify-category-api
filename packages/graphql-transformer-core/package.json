{
  "name": "graphql-transformer-core",
  "version": "8.2.4",
  "description": "A framework to transform from GraphQL SDL to AWS cloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-category-api.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "build": "rimraf ./lib ./tsconfig.tsbuildinfo && tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib",
    "extract-api": "ts-node ../../scripts/extract-api.ts"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^10.3.0",
    "graphql": "^15.5.0",
    "graphql-transformer-common": "4.27.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@aws-amplify/amplify-cli-core": "^4.2.10",
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^18.17.0",
    "mock-fs": "5.2.0",
    "@types/mock-fs": "4.13.4"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true,
    "coverageProvider": "v8",
    "coverageThreshold": {
      "global": {
        "branches": 67,
        "functions": 33,
        "lines": 48
      }
    },
    "coverageReporters": [
      "clover",
      "text"
    ]
  }
}
