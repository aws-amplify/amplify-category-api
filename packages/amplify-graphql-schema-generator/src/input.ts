import { print, InputObjectTypeDefinitionNode, DocumentNode } from 'graphql';
import { ImportedRDSType } from '@aws-amplify/graphql-transformer-core';

type AmplifyInputEntry = {
  name: string;
  type: string;
  default: string | number;
  comment?: string | undefined;
};

const getGlobalAmplifyInputEntries = (
  dataSourceType = ImportedRDSType.MYSQL,
  includeAuthRule = true,
  authDocLink?: string,
): AmplifyInputEntry[] => {
  const inputs: AmplifyInputEntry[] = [
    {
      name: 'engine',
      type: 'String',
      default: dataSourceType,
    },
  ];

  if (includeAuthRule) {
    inputs.push({
      name: 'globalAuthRule',
      type: 'AuthRule',
      default: '{ allow: public }',
      comment:
        'This "input" configures a global authorization rule to enable public access to all models in this schema.' +
        (!!authDocLink ? ` Learn more about authorization rules here:${authDocLink}` : ''),
    });
  }
  return inputs;
};

export const constructDefaultGlobalAmplifyInput = (
  dataSourceType: ImportedRDSType,
  includeAuthRule = true,
  authDocLink?: string,
): string => {
  const inputs = getGlobalAmplifyInputEntries(dataSourceType, includeAuthRule, authDocLink);
  const inputsString = inputs.reduce(
    (acc: string, input): string =>
      acc +
      ` ${input.name}: ${input.type} = ${input.type === 'String' ? '"' + input.default + '"' : input.default} ${
        input.comment ? '# ' + input.comment : ''
      } \n`,
    '',
  );
  return `input AMPLIFY {\n${inputsString}}\n`;
};

export const readRDSGlobalAmplifyInput = (schemaDocument: DocumentNode | undefined): InputObjectTypeDefinitionNode | undefined => {
  if (!schemaDocument) {
    return;
  }
  const inputNode = schemaDocument.definitions.find(
    (definition) => definition.kind === 'InputObjectTypeDefinition' && definition.name && definition.name.value === 'AMPLIFY',
  );

  if (inputNode) {
    return inputNode as InputObjectTypeDefinitionNode;
  }
};

export const constructRDSGlobalAmplifyInput = (config: any, schemaDocument: DocumentNode | undefined): string => {
  const existingInputNode: any = readRDSGlobalAmplifyInput(schemaDocument);
  if (existingInputNode && existingInputNode?.fields && existingInputNode?.fields?.length > 0) {
    const expectedInputs = getGlobalAmplifyInputEntries(ImportedRDSType.MYSQL).map((item) => item.name);
    expectedInputs.forEach((input) => {
      const inputNodeField = existingInputNode?.fields?.find((field: any) => field?.name?.value === input);
      if (inputNodeField && config[input]) {
        inputNodeField['defaultValue']['value'] = config[input];
      }
    });
    return print(existingInputNode);
  } else {
    const engine = config['engine'] || ImportedRDSType.MYSQL;
    return constructDefaultGlobalAmplifyInput(engine, false);
  }
};
