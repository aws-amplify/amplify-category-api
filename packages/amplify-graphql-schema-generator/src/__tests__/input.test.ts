import { ImportedRDSType } from '@aws-amplify/graphql-transformer-core';
import { constructDefaultGlobalAmplifyInput, readRDSGlobalAmplifyInput, constructRDSGlobalAmplifyInput } from '../input';
import { parse } from 'graphql';

jest.mock('fs-extra', () => ({
  readFileSync: jest.fn(),
  existsSync: jest.fn(),
}));

jest.mock('@aws-amplify/amplify-cli-core', () => {
  const original = jest.requireActual('@aws-amplify/amplify-cli-core');
  return {
    ...original,
    ApiCategoryFacade: {
      getTransformerVersion: jest.fn().mockResolvedValue(2),
    },
  };
});

describe('Amplify Input read/write from schema', () => {
  afterAll(() => {
    jest.resetAllMocks();
  });

  it('constructs valid default input parameters for MySQL datasource with global auth rule', async () => {
    const authDocLink = 'https://docs.amplify.aws/cli/graphql/authorization-rules';
    const expectedGraphQLInputString = `input AMPLIFY {
      engine: String = \"mysql\"
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:${authDocLink}
    }`;
    const constructedInputString = await constructDefaultGlobalAmplifyInput(ImportedRDSType.MYSQL, true, authDocLink);
    expect(constructedInputString?.replace(/\s/g, '')).toEqual(expectedGraphQLInputString.replace(/\s/g, ''));
  });

  it('reads global Amplify input type in the schema', async () => {
    const mockValidInputs = {
      engine: 'mysql',
      host: 'mockdatabase.rds.amazonaws.com',
      port: '1010',
      database: 'mockdatabase',
    };

    const mockInputSchema = `input AMPLIFY {
      engine: String = \"${mockValidInputs.engine}\"  
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:https://docs.amplify.aws/cli/graphql/authorization-rules 
    }`;

    const readInputNode = await readRDSGlobalAmplifyInput(parse(mockInputSchema));
    expect(readInputNode).toMatchSnapshot();
  });

  it('constructs the global Amplify input from given config', async () => {
    const mockValidInputs = {
      engine: 'mysql',
      host: 'mockdatabase.rds.amazonaws.com',
      port: '1010',
      database: 'mockdatabase',
    };

    const mockInputSchema = `input AMPLIFY {
      engine: String = \"${mockValidInputs.engine}\" 
      globalAuthRule: AuthRule = { allow: public } # This "input" configures a global authorization rule to enable public access to all models in this schema. Learn more about authorization rules here:https://docs.amplify.aws/cli/graphql/authorization-rules 
    }`;

    const userInputs = {
      host: 'mockdatabase.rds.amazonaws.com',
      port: 1010,
      database: 'mockdatabase',
    };

    const constructedInputDefinition = await constructRDSGlobalAmplifyInput(userInputs, parse(mockInputSchema));
    expect(constructedInputDefinition).toMatchSnapshot();
  });
});
