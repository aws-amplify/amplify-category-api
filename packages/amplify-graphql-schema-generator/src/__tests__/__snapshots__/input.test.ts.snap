// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amplify Input read/write from schema constructs the global Amplify input from given config 1`] = `
"input AMPLIFY {
  engine: String = \\"mysql\\"
  globalAuthRule: AuthRule = {allow: public}
}"
`;

exports[`Amplify Input read/write from schema reads global Amplify input type in the schema 1`] = `
Object {
  "description": undefined,
  "directives": Array [],
  "fields": Array [
    Object {
      "defaultValue": Object {
        "block": false,
        "kind": "StringValue",
        "loc": Object {
          "end": 46,
          "start": 39,
        },
        "value": "mysql",
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 46,
        "start": 22,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 28,
          "start": 22,
        },
        "value": "engine",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 36,
          "start": 30,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 36,
            "start": 30,
          },
          "value": "String",
        },
      },
    },
    Object {
      "defaultValue": Object {
        "fields": Array [
          Object {
            "kind": "ObjectField",
            "loc": Object {
              "end": 97,
              "start": 84,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 89,
                "start": 84,
              },
              "value": "allow",
            },
            "value": Object {
              "kind": "EnumValue",
              "loc": Object {
                "end": 97,
                "start": 91,
              },
              "value": "public",
            },
          },
        ],
        "kind": "ObjectValue",
        "loc": Object {
          "end": 99,
          "start": 82,
        },
      },
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "loc": Object {
        "end": 99,
        "start": 55,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 69,
          "start": 55,
        },
        "value": "globalAuthRule",
      },
      "type": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 79,
          "start": 71,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 79,
            "start": 71,
          },
          "value": "AuthRule",
        },
      },
    },
  ],
  "kind": "InputObjectTypeDefinition",
  "loc": Object {
    "end": 313,
    "start": 0,
  },
  "name": Object {
    "kind": "Name",
    "loc": Object {
      "end": 13,
      "start": 6,
    },
    "value": "AMPLIFY",
  },
}
`;
